packages/amplify-e2e-tests/src/__tests__/version-gating.test.ts

```diff
@@ -0,0 +1,232 @@
+      updateVersionInPackageJson('100.2.0', '100.2.0');
+
+      await push();
+
+      await expectStackMetadata('100.2.0', '100.2.0');
+
+      // Set CLI version to 100.3.0, Minimum CLI version to 100.2.0, verify push succeeds
+      updateVersionInPackageJson('100.3.0', '100.2.0');
+
+      await push();
+
+      await expectStackMetadata('100.3.0', '100.2.0');
+
+      // Set CLI version back to 100.2.0, Minimum CLI version to 100.2.0, verify push succeeds as
+      // minimum deploy version requirement has met
+      updateVersionInPackageJson('100.2.0', '100.2.0');
+
+      await push();
+
+      await expectStackMetadata('100.2.0', '100.2.0');
+    } finally {
+      // restore original package.json
+      fs.writeFileSync(packageJsonPath, originalPackageJsonContent);
+    }
+  });
+
+  // Test helper functions
+  const push = async () => {
+    await amplifyPushForceWithYes(projRoot);
+  };
+
+  const pushFail = async () => {
+    await amplifyPushForceWithVersionGatingOutput(projRoot);
+  };
+
+  const updateVersionInPackageJson = (cliVersion: string | undefined, minimumCLIVersion: string | undefined) => {
+    if (cliVersion) {
+      packageJson.version = cliVersion;
+    } else {
+      packageJson.version = baseCLIVersion;
+    }
+
+    if (minimumCLIVersion) {
+      packageJson.amplifyCLIConfiguration.minimumCLIVersion = minimumCLIVersion;
+    } else {
+      packageJson.amplifyCLIConfiguration.minimumCLIVersion = baseMinimumCLIVersion;
+    }
+
+    fs.writeFileSync(packageJsonPath, JSONUtilities.stringify(packageJson));
+  };
+
+  const expectStackMetadata = async (deployedByCLIVersion: string, minimumCompatibleCLIVersion: string) => {
+    const templateSummary = await cfnClient
+      .getTemplateSummary({
+        StackName: rootStackName,
+      })
+      .promise();
+
+    const metadataValue = _.get(templateSummary, ['Metadata']) || '{}';
+    const metadata = JSONUtilities.parse(metadataValue);
+    const versionGatingMetadata: VersionGatingMetadata = _.get(metadata, ['AmplifyCLI']);
+
+    expect(versionGatingMetadata.DeployedByCLIVersion).toBe(deployedByCLIVersion);
+    expect(versionGatingMetadata.MinimumCompatibleCLIVersion).toBe(minimumCompatibleCLIVersion);
+  };
+});
+
+const resolveRealCLIPath = (): string => {
+  const cliPath = getCLIPath(false);
+  const cliResolvedPath = which.sync(cliPath);
+
+  return fs.realpathSync(cliResolvedPath);
+};
+
+const isPackagedAmplifyInPath = (): boolean => {
+  const shebang = '#!/usr/bin/env node';
+  const isWin = process.platform.startsWith('win');
+
+  const cliRealPath = resolveRealCLIPath();
+
+  if (isWin) {
+    if (cliRealPath.endsWith('.cmd')) {
+      return false;
+    } else if (cliRealPath.endsWith('.exe')) {
+      return true;
+    }
+  }
+
+  // As *nix have no file extensions, read into the file and look for the shebang
+  let fileDescriptor: number | undefined;
+
+  try {
+    fileDescriptor = fs.openSync(cliRealPath, 'r');
```

Comment:
This is probably more efficient, but you could simplify the code by reading the entire file with `readFileSync()` and checking the start of the file. Up to you.

Reply:
Yes, I aimed for more efficient code